### ISSUE TITLE
Enhancement: Support for BigQuery SQL

### ISSUE THREAD

**Issue by barrywhart** [alanmcruickshank/sqlfluff#59]
Our data science projects use Google BigQuery. It will be good if this is supported. It's pretty close to standard SQL, I think.

**Comment by barrywhart** [comment]
As I encounter things that are not being parsed correctly, I'll try and determine if they are "standard" SQL or BigQuery-specific syntax. If standard, I'll create separate tickets. If they are BigQuery, would you prefer I add them as comments to this ticket or also create separate tickets?

**Comment by alanmcruickshank** [comment]
Can you post a couple of example queries here that currently fail? I'll assume that they should parse as standard sql until I find something in the syntax that conflicts. If we do find a conflict, then it will be a great excuse to test having multiple dialects, which I haven't found necessary yet.

**Comment by barrywhart** [comment]
Sure, I'll provide some examples. I have output for one whole project of ours. I plan to make a pass through it, fixing issues in our SQL and sharing any issues with you. Hopefully today or tomorrow.

**Comment by barrywhart** [comment]
Failing query:
```
SELECT
    user_id
FROM
    `project.dataset.user_age_labels`
```

This is the output:
```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   3 | P:   1 | ???? | Found unparsable segment @L003P001: 'FROM\n    `project.da...'
```

It seems the backquote characters are the issue here. If I remove them, the file is processed with no issues. In many cases (e.g. if the project name contains a `-`), BigQuery SQL *requires* the backquotes.

**Comment by barrywhart** [comment]
Context: BigQuery SQL supports [array columns](https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays). Square bracket subscripts in queries are breaking. This is a standard SQL feature, e.g.:
* Postgres: https://www.postgresql.org/docs/9.4/arrays.html
* DB2: https://www.ibm.com/support/knowledgecenter/en/SSEPEK_11.0.0/apsg/src/tpc/db2z_arrays.html

Failing query:
```
SELECT
  user_id,
  list_id,
  (count_18_24 * bits[OFFSET(0)] + count_25_34 * bits[OFFSET(1)] +
   count_35_44 * bits[OFFSET(2)] + count_45_54 * bits[OFFSET(3)] +
   count_55_64 * bits[OFFSET(4)] + count_65_plus * bits[OFFSET(5)]) / audience_size AS relative_abundance
FROM
    gcp_project.dataset.audience_counts_gender_age
```

This is the output:
```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   4 | P:  22 | ???? | Unable to lex characters: ''[OFFSET(0)'...'
```

**Comment by barrywhart** [comment]
These are the most common issues I see in our code. I'm still working through the output and will add more examples as I encounter them. From skimming the output, my intuition is that there are ~5-8 issues that, if addressed, would allow BigQuery to work pretty well for us.

I can potentially do some of this work, but I want to do a first pass through all the output first.

**Comment by barrywhart** [comment]
I have finished a first pass through all the output for the project I'm using to test SQLFluff. I believe if the items on this issue are addressed along with the other ones that I created separate tickets for, BigQuery will be working well enough for our project, and probably for SQLFluff to claim at least basic BigQuery compatibility.

Let me know if I can help addressing some of the issues. Once we've addressed some of them, I will be happy to test new versions to see if I notice any other issues.
